### PR TITLE
run tearDown() if TestCase fails with PHP error [closes #254]

### LINKED ISSUE #254: tearDown not executed when test fails with error or exception
I suppose this is a bug - this tidy method should be executed no matter if the test succeeds or fails. Actually tearDown works if the test fails on assertion, but not on error/exception.

//Edit: it works on exception, but not on any runtime error.

Please confirm it is a bug, I can fix and pullrequest it.

### PR DESCRIPTION
**Pull request by richard-ejem**

### CODE REVIEW

File: src/Framework/TestCase.php [hunk 1]
```diff
@@ -107,8 +107,12 @@ public function runTest($method, array $args = NULL)
 
 		foreach ($data as $params) {
 			try {
+				$me = $this; // PHP 5.3 compatibility
+				$lastHandler = set_error_handler(function() use ($me, &$lastHandler) {
+					$me->failsafeTearDown();
```
Comment (milo) [inline src/Framework/TestCase.php]: Restore handler here to prevent recursion.

File: src/Framework/TestCase.php [hunk 2]
```diff
@@ -174,6 +179,10 @@ protected function tearDown()
 	{
 	}
 
+	/** @internal */
+	public function failsafeTearDown() {
```
Comment (milo) [inline src/Framework/TestCase.php]: Don't add such method for now.

File: src/Framework/TestCase.php [hunk 3]
```diff
@@ -174,6 +179,10 @@ protected function tearDown()
...
+		$this->tearDown();
```
Comment (milo) [inline src/Framework/TestCase.php]: What about error in `tearDown()`. How to handle it?

File: src/Framework/TestCase.php [hunk 4]
```diff
@@ -107,8 +107,12 @@ public function runTest($method, array $args = NULL)
 
 		foreach ($data as $params) {
 			try {
+				$me = $this; // PHP 5.3 compatibility
+				$lastHandler = set_error_handler(function() use ($me, &$lastHandler) {
+					$me->failsafeTearDown();
+					return ($lastHandler === NULL) ? FALSE : $lastHandler();
```
Comment (milo) [inline src/Framework/TestCase.php]: Missing error handler arguments.

File: src/Framework/TestCase.php (hunk 3 above)
Comment (richard-ejem) [inline src/Framework/TestCase.php]: @milo well, if tearDown() fails, there is nothing more we can do (point of this pull request is to execute tearDown if setUp or test* fails).

File: src/Framework/TestCase.php (hunk 2 above)
Comment (richard-ejem) [inline src/Framework/TestCase.php]: The only point of this method is to make the stuff PHP 5.3 compatible - make possible to call tearDown from a callback, as PHP 5.3 doesn't see "$this" context from anonymous function.

File: src/Framework/TestCase.php (hunk 4 above)
Comment (richard-ejem) [inline src/Framework/TestCase.php]: thanks, fixed now

File: src/Framework/TestCase.php (hunk 1 above)
Comment (richard-ejem) [inline src/Framework/TestCase.php]: done ;)

File: src/Framework/TestCase.php [hunk 5]
```diff
@@ -174,6 +180,10 @@ protected function tearDown()
 	{
 	}
 
+	/** @internal */
+	public function failsafeTearDown() {
```
Comment (milo) [inline src/Framework/TestCase.php]: Name it as `_tearDown()` and fix coding style. Double empty line between methods, one empty line after the last one, curly bracket on separated line.

File: src/Framework/TestCase.php [hunk 6]
```diff
@@ -107,8 +107,13 @@ public function runTest($method, array $args = NULL)
 
 		foreach ($data as $params) {
 			try {
+				$me = $this; // PHP 5.3 compatibility
+				$lastHandler = set_error_handler(function() use ($me, &$lastHandler) {
+					restore_error_handler();
+					$me->failsafeTearDown();
+					return ($lastHandler === NULL) ? FALSE : call_user_func_array($lastHandler, func_get_args());
+				});
 				$this->setUp();
```
Comment (milo) [inline src/Framework/TestCase.php]: Move `setUp()` call before error handler, exceptions aren't catched too.

File: src/Framework/TestCase.php [hunk 7]
```diff
@@ -107,8 +107,13 @@ public function runTest($method, array $args = NULL)
 
 		foreach ($data as $params) {
 			try {
+				$me = $this; // PHP 5.3 compatibility
+				$lastHandler = set_error_handler(function() use ($me, &$lastHandler) {
+					restore_error_handler();
+					$me->failsafeTearDown();
+					return ($lastHandler === NULL) ? FALSE : call_user_func_array($lastHandler, func_get_args());
+				});
```
Comment (milo) [inline src/Framework/TestCase.php]: Fix coding style. Space after `()`, space after `&`.

**Comment by milo** [comment]
What about
```php
	/** @internal */
	public function _errorHandler()
	{
		restore_error_handler();
		@$this->tearDown(); // ignore, error from testMethod() has higher precedence

		return $this->previousHandler
			? call_user_func_array($this->previousHandler, func_get_args())
			: FALSE;
	}
```

File: src/Framework/TestCase.php (hunk 7 above)
Comment (pavelkouril) [inline src/Framework/TestCase.php]: Also, shouldn't it be `function ()`?

**Comment by richard-ejem** [comment]
@milo thanks for suggestion, I like your _errorHandler alternative :) fixed in my code

**Comment by milo** [comment]
Solved by d25b67216a30a723db040ebcfd2c10e446f12218.

**Pull request closed without merging**